### ISSUE TITLE
Shuffles always seem clumped.

### ISSUE THREAD

**Issue by 4ist-Gump** [Cockatrice/Cockatrice#2945]
<b>OS:</b>
Win10v1709

<b>Cockatrice version:</b>
2.4

__________

I'm just getting into using Cockatrice, so take this with a grain of salt, but all the games I've run and single-player testing I've done appears to always have clumps in the library; regardless how many times I shuffle or not.  When I play tabletop or finish building a deck, I reassemble the deck so that there is no clumps by default (separating like-cards in stacks of 4 and then every 3rd card a land) and then shuffling several times.

I thought this was the norm, maybe I'm just weird?

At any rate, the game seems to stack the cards from how it sees them prior to a game beginning and shuffling from there... which creates clumps every-single-time.  I feel I don't get a real-world sense of the deck until I go through this process a dozen times.

Having tried a few of the MTGO UI, this is definitely my favorite.  You guys have done an awesome job so far.

**Comment by tooomm** [comment]
>I reassemble the deck so that there is no clumps by default (separating like-cards in stacks of 4 and then every 3rd card a land) and then shuffling several times.

That is not shuffling, that's called mana weaving (or cheating). 😆 

Generally there is the saying that you can't perfectly shuffle in real life, while the computer tries to do that and most likely will perform way better than you ever could in person. That's the reason why people feel the difference and think computer shufflers are not workingor broken.

I think we use a commonly used algorithm for shuffling which should produce random results.

Can somebody come up with a proper explanation, so we can add it to our FAQ actually?

**Comment by 4ist-Gump** [comment]
I didn't realize I was cheating at tabletop.  =X   I just assumed everyone starts their series of playing from a neutral/default point.

So then, the randomness will likely always result in some clumping, the real question is: to what degree will the clumping be?

**Comment by Daenyth** [comment]
We use the well-known Fischer-Yates shuffling algorithm, and the server has a chi-square test that shows the shuffling to be evenly distributed - meaning that any permutation of cards in the deck is equally likely. The degree of clumping then can be calculated using statistics based on your deck list.

tldr; the shuffler is as random as you can get on a computer, if you can notice differences between in person and on cockatrice then your in person shuffling isn't actually fully randomized.
